deps/src/caller.cpp

```diff
@@ -0,0 +1,340 @@
+static jl_value_t * jl_singular_number_type;
+static jl_value_t * jl_singular_poly_type;
+static jl_value_t * jl_singular_ring_type;
+static jl_value_t * jl_singular_ideal_type;
+static jl_value_t * jl_singular_matrix_type;
+static jl_value_t * jl_singular_bigint_type;
+static jl_value_t * jl_singular_bigintmat_type;
+static jl_value_t * jl_singular_map_type;
+static jl_value_t * jl_singular_resolution_type;
+
+static jl_value_t * get_type_mapper()
+{
+    std::vector<std::pair<int, std::string>> types = {
+        std::pair<int, std::string>(BIGINT_CMD, "BIGINT_CMD"),
+        std::pair<int, std::string>(NUMBER_CMD, "NUMBER_CMD"),
+        std::pair<int, std::string>(RING_CMD, "RING_CMD"),
+        std::pair<int, std::string>(POLY_CMD, "POLY_CMD"),
+        std::pair<int, std::string>(IDEAL_CMD, "IDEAL_CMD"),
+        std::pair<int, std::string>(INT_CMD, "INT_CMD"),
+        std::pair<int, std::string>(STRING_CMD, "STRING_CMD"),
+        std::pair<int, std::string>(LIST_CMD, "LIST_CMD"),
+        std::pair<int, std::string>(INTMAT_CMD, "INTMAT_CMD"),
+        std::pair<int, std::string>(BIGINTMAT_CMD, "BIGINTMAT_CMD"),
+        std::pair<int, std::string>(MAP_CMD, "MAP_CMD"),
+        std::pair<int, std::string>(RESOLUTION_CMD, "RESOLUTION_CMD"),
+        std::pair<int, std::string>(INTVEC_CMD, "INTVEC_CMD")};
+
+    jl_array_t * return_array =
+        jl_alloc_array_1d(jl_array_any_type, types.size());
+
+    for (int i = 0; i < types.size(); i++) {
+        jl_array_t * current_return = jl_alloc_array_1d(jl_array_any_type, 2);
+        jl_arrayset(current_return, jl_box_int64(types[i].first), 0);
+        jl_arrayset(current_return,
+                    reinterpret_cast<jl_value_t *>(
+                        jl_symbol(types[i].second.c_str())),
+                    1);
+        jl_arrayset(return_array,
+                    reinterpret_cast<jl_value_t *>(current_return), i);
+    }
+    return reinterpret_cast<jl_value_t *>(return_array);
+}
+
+static void initialize_jl_c_types(jl_value_t* module_value)
+{
+    jl_module_t* module = reinterpret_cast<jl_module_t*>(module_value);
+    jl_int64_vector_type =
+        jl_apply_array_type((jl_value_t *)jl_int64_type, 1);
+    jl_int64_matrix_type =
+        jl_apply_array_type((jl_value_t *)jl_int64_type, 2);
+    jl_singular_number_type = jl_get_global(module,jl_symbol("number"));
+    jl_singular_poly_type = jl_get_global(module,jl_symbol("poly"));
+    jl_singular_ring_type = jl_get_global(module,jl_symbol("ring"));
+    jl_singular_ideal_type = jl_get_global(module,jl_symbol("ideal"));
+    jl_singular_matrix_type = jl_get_global(module,jl_symbol("ip_smatrix"));
+    jl_singular_bigint_type = jl_get_global(module,jl_symbol("__mpz_struct"));
+    jl_singular_bigintmat_type = jl_get_global(module,jl_symbol("bigintmat"));
+    jl_singular_bigintmat_type = jl_get_global(module,jl_symbol("sip_smap"));
+    jl_singular_resolution_type = jl_get_global(module,jl_symbol("resolvente"));
```

Comment:
Totally unrelated to this PR, but: huh, what is a "resolvente"? Or is that "denglish" / "germanglish" for "resolvent"? But even then, I never heard the term "resolvent" used as an alias for "resolution" in this context (googling reveals such a use in logic, but not in commutative algebra). But of course I am hardly an expert, so this is not exactly a surprise; I'd still be curious to learn about a reference or two where this word is used like this.


Uhh, I see that this is defined in Singular, and also has been exposed under the same name in `Singular.jl`.